File: android-im-ui/src/test/kotlin/com/flare/im/ui/ConversationActionSheetTest.kt
package com.flare.im.ui

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ConversationActionSheetTest {
    private val base = FlareConversationActionSnapshot("c1", "设计评审")
    private val all = FlareConversationActionCapabilities(pin = true, mute = true, markRead = true, archive = true, delete = true, hide = true)
    private fun ids(c: FlareConversationActionSnapshot, caps: FlareConversationActionCapabilities) =
        conversationActions(c, caps).map { it.action }

    @Test fun nothingWithoutCapabilities() {
        assertEquals(emptyList(), ids(base, FlareConversationActionCapabilities()))
    }

    @Test fun eachCapabilityRevealsItsAction() {
        assertEquals(listOf(FlareConversationAction.Pin), ids(base, FlareConversationActionCapabilities(pin = true)))
        assertEquals(listOf(FlareConversationAction.Mute), ids(base, FlareConversationActionCapabilities(mute = true)))
        assertEquals(listOf(FlareConversationAction.Archive), ids(base, FlareConversationActionCapabilities(archive = true)))
        assertEquals(listOf(FlareConversationAction.Hide), ids(base, FlareConversationActionCapabilities(hide = true)))
        assertEquals(listOf(FlareConversationAction.Delete), ids(base, FlareConversationActionCapabilities(delete = true)))
    }

    @Test fun invertsByState() {
        assertEquals(listOf(FlareConversationAction.Unpin), ids(base.copy(pinned = true), FlareConversationActionCapabilities(pin = true)))
        assertEquals(listOf(FlareConversationAction.Unmute), ids(base.copy(muted = true), FlareConversationActionCapabilities(mute = true)))
        assertEquals(listOf(FlareConversationAction.Unarchive), ids(base.copy(archived = true), FlareConversationActionCapabilities(archive = true)))
    }

    @Test fun markReadOnlyWithUnread() {
        assertEquals(emptyList(), ids(base, FlareConversationActionCapabilities(markRead = true)))
        assertEquals(listOf(FlareConversationAction.MarkRead), ids(base.copy(unreadCount = 3), FlareConversationActionCapabilities(markRead = true)))
    }

    @Test fun orderAndDangerGroup() {
        val entries = conversationActions(base.copy(unreadCount = 2), all)
        assertEquals(
            listOf(FlareConversationAction.Pin, FlareConversationAction.Mute, FlareConversationAction.MarkRead,
                FlareConversationAction.Archive, FlareConversationAction.Hide, FlareConversationAction.Delete),
            entries.map { it.action },
        )
        assertEquals(listOf(FlareConversationAction.Delete), entries.filter { it.danger }.map { it.action })
        assertTrue(entries.last().danger)
    }
}

File: flutter-im-ui/test/conversation_action_sheet_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flare_im_ui/src/components/flare_conversation_action_sheet.dart';

void main() {
  const base = FlareConversationActionSnapshot(id: 'c1', title: '设计评审');
  const all = FlareConversationActionCapabilities(
      pin: true, mute: true, markRead: true, archive: true, hide: true, delete: true);
  List<FlareConversationAction> ids(
          FlareConversationActionSnapshot c, FlareConversationActionCapabilities caps) =>
      conversationActions(c, caps).map((e) => e.action).toList();

  group('conversationActions', () {
    test('nothing without capabilities', () {
      expect(ids(base, const FlareConversationActionCapabilities()), isEmpty);
    });
    test('each capability reveals exactly its action', () {
      expect(ids(base, const FlareConversationActionCapabilities(pin: true)),
          [FlareConversationAction.pin]);
      expect(ids(base, const FlareConversationActionCapabilities(mute: true)),
          [FlareConversationAction.mute]);
      expect(ids(base, const FlareConversationActionCapabilities(archive: true)),
          [FlareConversationAction.archive]);
      expect(ids(base, const FlareConversationActionCapabilities(hide: true)),
          [FlareConversationAction.hide]);
      expect(ids(base, const FlareConversationActionCapabilities(delete: true)),
          [FlareConversationAction.delete]);
    });
    test('inverts pin/mute/archive by state', () {
      expect(
          ids(const FlareConversationActionSnapshot(id: 'c', title: 't', pinned: true),
              const FlareConversationActionCapabilities(pin: true)),
          [FlareConversationAction.unpin]);
      expect(
          ids(const FlareConversationActionSnapshot(id: 'c', title: 't', muted: true),
              const FlareConversationActionCapabilities(mute: true)),
          [FlareConversationAction.unmute]);
      expect(
          ids(const FlareConversationActionSnapshot(id: 'c', title: 't', archived: true),
              const FlareConversationActionCapabilities(archive: true)),
          [FlareConversationAction.unarchive]);
    });
    test('markRead only when unread > 0', () {
      expect(ids(base, const FlareConversationActionCapabilities(markRead: true)), isEmpty);
      expect(
          ids(const FlareConversationActionSnapshot(id: 'c', title: 't', unreadCount: 3),
              const FlareConversationActionCapabilities(markRead: true)),
          [FlareConversationAction.markRead]);
    });
    test('stable order, delete last and danger', () {
      final entries = conversationActions(
          const FlareConversationActionSnapshot(id: 'c', title: 't', unreadCount: 2), all);
      expect(entries.map((e) => e.action).toList(), [
        FlareConversationAction.pin,
        FlareConversationAction.mute,
        FlareConversationAction.markRead,
        FlareConversationAction.archive,
        FlareConversationAction.hide,
        FlareConversationAction.delete,
      ]);
      expect(entries.where((e) => e.danger).map((e) => e.action), [FlareConversationAction.delete]);
    });
  });

  testWidgets('renders labels, emits action, busy disables', (tester) async {
    final received = <String>[];
    Widget build({bool busy = false}) => MaterialApp(
          home: Scaffold(
            body: FlareConversationActionSheet(
              conversation: const FlareConversationActionSnapshot(
                  id: 'c1', title: '设计评审', pinned: true, unreadCount: 1),
              capabilities: all,
              busy: busy,
              onAction: (id, a) => received.add('$id/${a.name}'),
            ),
          ),
        );
    await tester.pumpWidget(build());
    expect(find.text('取消置顶'), findsOneWidget);
    expect(find.text('标为已读'), findsOneWidget);
    expect(find.text('删除'), findsOneWidget);
    expect(find.text('置顶'), findsNothing);

    await tester.tap(find.text('删除'));
    await tester.pump();
    expect(received, ['c1/delete']);

    await tester.pumpWidget(build(busy: true));
    await tester.tap(find.text('取消置顶'));
    await tester.pump();
    expect(received, ['c1/delete']);
    expect(tester.takeException(), isNull);
  });

  testWidgets('empty capabilities show the empty text, not a blank menu', (tester) async {
    await tester.pumpWidget(const MaterialApp(
        home: Scaffold(body: FlareConversationActionSheet(conversation: base))));
    expect(find.text('暂无可用操作'), findsOneWidget);
  });
}
